### PR TITLE
New uncaught interface

### PR DESCRIPTION
**Pull request by cristiancavalli**
This addresses ticket #16. This commit builds off pull request #22.

### CODE REVIEW

**Comment by cristiancavalli** [comment]
@matthewloring @ofrobots PTAL

**Comment by ofrobots** [comment]
This PR is doing two things at the same time. I would prefer separate commits and/or PRs for distinct things. 
* I want to be able to review the change and test for #16 independently of the change for #29.
* We need a bit of discussion before implementing a fix for #29.
* Commit log: 50 char abstract. We follow the commit guide from node-core: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#step-3-commit.

**Comment by cristiancavalli** [comment]
@ofrobots pull request changed to focus only on #16

**Comment by cristiancavalli** [comment]
build failed on a coveralls error
Bad response: 422 {"message":"Couldn't find a repository matching this job.","error":true}

restarting..

File: lib/interfaces/uncaught.js [hunk 1]
```diff
@@ -47,14 +46,7 @@ function uncaughtExceptionHandler(client, config, err) {
...
 
-  if (config.onUncaughtException === uncaughtHandlingOptions.report) {
-
-    client.sendError(em);
-  } else if (config.onUncaughtException ===
-             uncaughtHandlingOptions.reportAndExit) {
-
-    client.sendError(em, handleProcessExit);
-  }
+  client.sendError(em, handleProcessExit);
```
Comment (ofrobots) [inline lib/interfaces/uncaught.js]: You no longer need `config` in this function, so you can remove it from the signature.

File: lib/configuration.js [hunk 2]
```diff
@@ -56,7 +54,7 @@ function extractServiceContext(initConfiguration) {
   };
 
   if (has(initConfiguration, 'serviceContext') &&
-      isObject(initConfiguration.serviceContext)) {
+      isPlainObject(initConfiguration.serviceContext)) {
```
Comment (ofrobots) [inline lib/configuration.js]: Is this change logically related to how the uncaught exception configuration works?

File: lib/configuration.js [hunk 3]
```diff
@@ -76,9 +74,8 @@ function extractServiceContext(initConfiguration) {
  * @typedef NormalizedConfigurationVariables
  * @type Object
  * @property {String} projectId - the project id of the hosting application
- * @property {uncaughtHandlingEnum}
- *  onUncaughtException=uncaughtHandlingEnum.ignore - the flag denoting how to
- *  handle uncaught exceptions
+ * @property {Boolean} reportUncaughtExceptions=true - the flag denoting
+ *  whether or not to report uncaught errors to the API
```
Comment (ofrobots) [inline lib/configuration.js]: "whether or not to catch and report uncaught"

File: lib/configuration.js [hunk 4]
```diff
@@ -149,16 +145,20 @@ function gatherConfiguration(initConfiguration) {
         'An invalid configuration was given to the Stackdriver Error Library');
   }
 
+  if (!isString(projectId)) {
+     if (isString(env.GCLOUD_PROJECT) || isNumber(env.GCLOUD_PROJECT)) {
+
+      projectId = env.GCLOUD_PROJECT;
```
Comment (ofrobots) [inline lib/configuration.js]: This doesn't seem logically related to uncaught exceptions. Can this be moved to a PR of its own?

File: lib/configuration.js (hunk 2 above)
Comment (cristiancavalli) [inline lib/configuration.js]: @ofrobots it does not. It was just a quick fix to eliminate a false positive where a user could supply an array.

Comment (ofrobots) [inline lib/configuration.js]: sgtm

File: lib/configuration.js (hunk 3 above)
Comment (cristiancavalli) [inline lib/configuration.js]: fixed

File: lib/interfaces/uncaught.js (hunk 1 above)
Comment (cristiancavalli) [inline lib/interfaces/uncaught.js]: fixed

File: lib/configuration.js (hunk 4 above)
Comment (cristiancavalli) [inline lib/configuration.js]: @ofrobots sounds good, this change has been reversed

File: lib/configuration.js [hunk 5]
```diff
@@ -149,16 +145,17 @@ function gatherConfiguration(initConfiguration) {
...
-
   if (isString(env.GCLOUD_PROJECT) || isNumber(env.GCLOUD_PROJECT)) {
-
     projectId = env.GCLOUD_PROJECT;
   }
 
+  serviceContext = extractServiceContext(initConfiguration);
+
+
+
```
Comment (ofrobots) [inline lib/configuration.js]: nit: so much whitespace.

**Comment by ofrobots** [comment]
LGTM w/ nit.

**Comment by cristiancavalli** [comment]
merging after ci pass

**Pull request merged**